Content.Server/GameObjects/Components/BaseComputerUserInterfaceComponent.cs

```diff
@@ -0,0 +1,144 @@
+            UserInterfaceKey = key;
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            if (UserInterface != null)
+                UserInterface.OnReceiveMessage += OnReceiveUIMessageCallback;
+
+            Owner.EnsureComponent<PowerReceiverComponent>();
+        }
+
+        /// <summary>
+        /// Internal callback used to grab session and session attached entity before any more work is done.
+        /// This is so that sessionEntity is always available to checks up and down the line.
+        /// </summary>
+        private void OnReceiveUIMessageCallback(ServerBoundUserInterfaceMessage obj)
+        {
+            var session = obj.Session;
+            var sessionEntity = session.AttachedEntity;
+            if (sessionEntity == null)
+                return; // No session entity, so we're probably not able to touch this.
+            OnReceiveUnfilteredUserInterfaceMessage(obj, sessionEntity);
+        }
+
+        /// <summary>
+        /// Override this to handle messages from the UI before filtering them.
+        /// Calling base is necessary if you want this class to have any meaning.
+        /// </summary>
+        protected void OnReceiveUnfilteredUserInterfaceMessage(ServerBoundUserInterfaceMessage obj, IEntity sessionEntity)
+        {
+            // "Across all computers" "anti-cheats" ought to be put here or at some parent level (BaseDeviceUserInterfaceComponent?)
+            // Determine some facts about the session.
+            // Powered?
+            if (!Powered)
+            {
+                sessionEntity.PopupMessageCursor(Loc.GetString("base-computer-ui-component-not-powered"));
+                return; // Not powered, so this computer should probably do nothing.
+            }
+            // Can we interact?
+            if (!ActionBlockerSystem.CanInteract(sessionEntity))
+            {
+                sessionEntity.PopupMessageCursor(Loc.GetString("base-computer-ui-component-cannot-interact"));
+                return;
+            }
+            // Good to go!
+            OnReceiveUserInterfaceMessage(obj);
+        }
+
+        /// <summary>
+        /// Override this to handle messages from the UI.
+        /// Calling base is unnecessary.
+        /// These messages will automatically be blocked if the user shouldn't be able to access this computer, or if the computer has lost power.
+        /// </summary>
+        protected virtual void OnReceiveUserInterfaceMessage(ServerBoundUserInterfaceMessage obj)
+        {
+            // Nothing!
+        }
+
+        public override void HandleMessage(ComponentMessage message, IComponent? component)
+        {
+            base.HandleMessage(message, component);
+            switch (message)
+            {
+                case PowerChangedMessage powerChanged:
+                    PowerReceiverOnOnPowerStateChanged(powerChanged);
+                    break;
+            }
+        }
```

Comment:
Don't use component messages as they're marked as obsolete.

Reply:
This code was moved from it's previous place. Is there a replacement way to get the same functionality, and if so, what?

Reply:
Use an EntitySystem. Create a PowerStateChangedEvent that inherits from EntityEventArgs and raise it where PowerReceiverOnOnPowerStateChanged is sent from.

In the EntitySystem subscribe to the event and handle stuff through that.

Reply:
Does this still apply or does the later "so just fix the EnsureComponent part and this can be merged" apply?